### PR TITLE
validate access token

### PR DESCRIPTION
**Pull request by ZhuXiaoBing-cn**
* Add single tenant and multi-tenant verification
* Map scope in access token to authority

### CODE REVIEW

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAppRoleStatelessAuthenticationFilter.java [hunk 1]
```diff
@@ -107,4 +113,19 @@ private boolean alreadyAuthenticated() {
...
+                .map(scopes->scopes.split(" "))
+                .map(Arrays::stream)
+                .orElseGet(Stream::empty)
+                .filter(StringUtils::hasText)
+                .map(s -> new SimpleGrantedAuthority(SCOPE_PREFIX + s))
+                .collect(Collectors.toSet());
+
+        return Optional.of(scopeSimpleGrantedAuthoritySet)
+            .filter(r -> !r.isEmpty())
+            .orElse(DEFAULT_AUTHORITY_SET);
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAppRoleStatelessAuthenticationFilter.java:129]: We don't need this here, we are parsing scopes in this function, it's not reasonable to output a set of ROLE_xxx.

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAppRoleStatelessAuthenticationFilter.java [hunk 2]
```diff
@@ -65,10 +67,14 @@ protected void doFilterInternal(HttpServletRequest httpServletRequest,
         }
         try {
             final UserPrincipal userPrincipal = principalManager.buildUserPrincipal(aadIssuedBearerToken);
+            Set<SimpleGrantedAuthority> authorities = toSimpleGrantedAuthoritySet(userPrincipal);
+            if(userPrincipal.getClaim(AADTokenClaim.SCP) != null){
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAppRoleStatelessAuthenticationFilter.java]: Should we move this condition into the scopeToSimpleGrantedAuthoritySet?

Comment (saragluna) [reply sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAppRoleStatelessAuthenticationFilter.java]: The `authorities` returned by the first function could be a unmodifiableSet. If we try to add scopes here, an exception will be thrown.

Comment (saragluna) [reply sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAppRoleStatelessAuthenticationFilter.java]: Maybe we should combine these two functions together.

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADTokenClaim.java [hunk 3]
```diff
@@ -11,4 +11,5 @@
     public static final String NAME = "name";
     public static final String TID = "tid";
     public static final String ROLES = "roles";
+    public static final String SCP = "scp";
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADTokenClaim.java:14]: Is this the only possible scope claim name from AAD token? Could it be scope here?

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/UserPrincipalManager.java [hunk 4]
```diff
@@ -181,31 +185,56 @@ private static boolean isAADIssuer(String issuer) {
...
                         LOGGER.debug("Matched audience: [{}]", matchedAudience.get());
                     } else {
-                        throw new BadJWTException("Invalid token audience. Provided value " + claimsSet.getAudience()
-                            + "does not match neither client-id nor AppIdUri.");
+                        throw new BadJWTException(
+                            "Invalid token audience. Provided value " + claimsSet.getAudience()
+                                + "does not match neither client-id nor AppIdUri.");
                     }
                 }
+                if (aadAuthenticationProperties.getAllowedTenantIds().isEmpty() &&
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/UserPrincipalManager.java:229]: Please refactor these conditions, such as extracting local variables.

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/UserPrincipalManager.java [hunk 5]
```diff
@@ -181,31 +185,56 @@ private static boolean isAADIssuer(String issuer) {
             || issuer.startsWith(STS_CHINA_CLOUD_API_ISSUER);
     }
 
+    private boolean isAADTenant(String tenant) {
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/UserPrincipalManager.java:188]: This function name is confusing, I suppose all tenants are AAD tenant right?

**Review by saragluna — CHANGES_REQUESTED** [review]

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADTokenClaim.java (hunk 3 above)
Comment (ZhuXiaoBing-cn) [reply sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADTokenClaim.java:14]: yes,this is the only possible scp claim name from AAD token,I got it from the Microsoft docs.

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAppRoleStatelessAuthenticationFilter.java (hunk 1 above)
Comment (ZhuXiaoBing-cn) [reply sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAppRoleStatelessAuthenticationFilter.java:129]: yes,I'm going to modify it

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAppRoleStatelessAuthenticationFilter.java (hunk 2 above)
Comment (ZhuXiaoBing-cn) [reply sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAppRoleStatelessAuthenticationFilter.java]: The combine will look even better.

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/UserPrincipalManager.java (hunk 5 above)
Comment (ZhuXiaoBing-cn) [reply sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/UserPrincipalManager.java:188]: yes,I'm going to modify it

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/UserPrincipalManager.java (hunk 4 above)
Comment (ZhuXiaoBing-cn) [reply sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/UserPrincipalManager.java:229]: okay

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAppRoleStatelessAuthenticationFilter.java [hunk 6]
```diff
@@ -101,10 +103,22 @@ private boolean alreadyAuthenticated() {
                     .map(Collection::stream)
                     .orElseGet(Stream::empty)
                     .filter(StringUtils::hasText)
-                    .map(s -> new SimpleGrantedAuthority(ROLE_PREFIX + s))
+                    .map(role -> new SimpleGrantedAuthority(ROLE_PREFIX + role))
                     .collect(Collectors.toSet());
+        if (userPrincipal.getClaim(AADTokenClaim.SCP) != null) {
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAppRoleStatelessAuthenticationFilter.java:108]: `!= null` can be included in `Optional`.

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AzureADGraphClient.java [hunk 7]
```diff
@@ -12,6 +12,8 @@
 import com.microsoft.aad.msal4j.OnBehalfOfParameters;
 import com.microsoft.aad.msal4j.UserAssertion;
 import com.nimbusds.oauth2.sdk.http.HTTPResponse;
+import java.util.Arrays;
+import java.util.stream.Stream;
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AzureADGraphClient.java:16]: https://github.com/Azure/azure-sdk-for-java/pull/17325/files
After this(jialin's) PR merged, we will not use `AzureADGraphClient` any more. 

So my suggestion is to abandon this PR and learn jialin's PR. 

Please confirm that in jialin's PR, whether `scope/role` Authority work well with multiple resourceUri and multiple tenant. 🙏

**Review by chenrujun — CHANGES_REQUESTED** [review]

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AzureADGraphClient.java (hunk 7 above)
Comment (ZhuXiaoBing-cn) [reply sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AzureADGraphClient.java:16]: Okay,thank you for your suggestion,
`confirm that in jialin's PR, whether scope/role Authority work well with multiple resourceUri and multiple tenant. `
I will confirm it.

**Pull request closed without merging**